---
title: "`projectName`"
author: "Chandra Chilamakuri"
date: "`r format(Sys.time(), '%a %d %b %Y')`"
output:
  
  
  html_document:

    theme: cerulean
    highlight: espresso
    toc: yes
    toc_float: yes
    smooth_scroll: yes
    code_folding: hide
---

```{r setup, include=FALSE}
# load libraries
#suppressPackageStartupMessages(library(rnaseqRcode) )# Analysis R package
devtools::load_all()
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(assertthat))
# knitr global options
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning = FALSE, cache = FALSE)
```


```{r def_functions}
catHeader <- function(text = "", level = 3, is_tabset=FALSE) {
  
  if(is_tabset){
    cat(paste0("\n\n", 
               paste(rep("#", level), collapse = ""), 
               " ", text,' {.tabset}', "\n"))
    
  }else{
    cat(paste0("\n\n", 
               paste(rep("#", level), collapse = ""), 
               " ", text, "\n"))
  }
}

```

### Define veriables

```{r def_vars} 
# define variables
projectName <- 'Test project'
genome <- 'Mus_musculus'
quantOut <- 'data/quantOut'
tx2geneFile <- 'data/references/tx2gene.tsv'
gtfFile <- 'data/references/mmu.GRCm38.gtf'
contrastFile <- 'data/contrasts.csv'
design <- '~SampleGroup'
countsDir <- 'data/counts'
factorName <- 'SampleGroup'
DeOutDir <- 'data/DEAnalysis'
pValCutoff <- 0.05
genesToShow <- 'ESR1'

#################### LOAD DATA ######################################
# load sample metadata
s_sheet <- read_csv('data/samplesheet_corrected.csv', progress=FALSE, 
                   show_col_types = FALSE)

# load tx2gene data
tx2gene <- read_tsv(tx2geneFile, show_col_types = FALSE)

# GTF
#gtf <- loadGTF(gtfFile = gtfFile)
#saveRDS(object = gtf, file = 'data/gtf.rds')
gtf <- readRDS('data/gtf.rds')
# contrast table
contTab <- read_csv(contrastFile, show_col_types = FALSE)
```

## Sample information {.tabset}

### Sample metadata 

```{r sample_metadata}
groupNames <-  unique(s_sheet$SampleGroup)
noGroups <- length(groupNames)
colPals = brewer.pal.info[brewer.pal.info$colorblind, ] %>% 
  .[c('Dark2', 'Paired'),]

colVector = unlist(mapply(brewer.pal, colPals$maxcolors,
                           rownames(colPals)))

if (noGroups < length(colVector) ){
  groupColors <- sample(colVector, noGroups, replace = FALSE)
}else{
  groupColors <- sample(colVector, noGroups, replace = TRUE)
}

s_sheet %>% 
  arrange('SampleGroup') %>% 
  datatable(style='bootstrap') %>% 
  formatStyle(
    columns = 'SampleGroup',
    target = 'row',
    color= styleEqual(groupNames, groupColors)
  )
```


###  Replicates per sample group

```{r sampleGroupTab}
s_sheet %>% 
  select(SampleGroup) %>% 
  group_by(SampleGroup) %>% 
  summarise(Replicates=n()) %>% 
  datatable()
```

```{r load_data}
txi <- importSampleTxiAndSaveRds(s_sheet = s_sheet, quantOut = quantOut, 
                          quantTool = 'salmon', tx2gene = tx2gene )
rawCounts <- txi$counts
mode(rawCounts) <- 'integer'
rawCounts <- rawCounts[rowSums(rawCounts) > 0, ]
```

# QC 

##  QC on reads per sample {.tabset}   

### reads / sample table

```{r reads_per_sample_tab}
readsPerSample <- getReadCountsFromSalmonLogs(s_sheet = s_sheet, 
                                              quantOut = quantOut) %>%
  mutate(fragments = fragments / 1000000) %>% 
  mutate(fragments = round(fragments, digits = 2))
  

readsPerSample %>% 
  datatable(caption = 'Reads per sample (million)') 

```

### barplot: reads / sample 

```{r reads_per_sample_bp}
readsPerSampleBarPlot(readCounts = readsPerSample)
```

### boxplot: reads / sample group

```{r reads_per_group_boxplot}
readsPerGroupBoxplot(s_sheet = s_sheet, readCounts = readsPerSample)
```

## QC on raw count matrix  {.tabset}

### density plot: reads / sample group

```{r reads_per_sample_dens_plot}
countsDensityPlot(s_sheet = s_sheet, countsData = rawCounts)
```

### Ribosomal read fraction

```{r robo_fraction_plot}
riboFractionPlot(countsData = rawCounts, gtf = gtf)
```

### counts on gene biotype  

```{r gene_biotype_plot}
plotGeneBiotypes(countsData = rawCounts, gtf=gtf)
```

### Library complexity

```{r library_complexity_plot}
getLibraryComplexityPlot(countsData = rawCounts, s_sheet = s_sheet)
```


### Gene detection 

```{r gene_detec_plot}
getGeneDetctionPlot(countsData = rawCounts, s_sheet = s_sheet)
```



## QC: Unsupervised clustering {.tabset}

### PCA 

```{r PCA_plot}
trnCounts <- transformCounts(rawCounts = rawCounts, countsCutOff = 10, FUN = vst)
getPcaPlot(countsDat = trnCounts, s_sheet = s_sheet,pcaColFactor = 'SampleGroup')
```

### PC loadings plot 

```{r pc_loadings_plot}
getPcaLoadingsPlot(countsData=trnCounts, s_sheet = s_sheet,
                   genesToShow = NULL, gtf = gtf, topN = 10)
```


### Hierarchical clustering 

```{r hclust_plot}
hierarchicalClustPlot(countsDat=trnCounts, 
                      s_sheet = s_sheet, 
                      colorByCol = 'SampleGroup' )
```

### correlogram 

```{r correlogram_plot}
correlationPlot(countsMat = trnCounts, s_sheet = s_sheet)
```


# DE analysis

```{r DE_analysis}
design <- checkAndCreateCorrectDesignString(s_sheet = s_sheet, 
                                            design = design)

dds <- createDdsAndDESeq(txi=txi, s_sheet = s_sheet, design = design )

saveCounts(dds = dds, txi = txi, countsDir = countsDir, gtf = gtf )
```

## DESeq2 QC plots {.tabset}

### Normalisation plot 

```{r norm_plots}
normFactorsBoxplot(dds=dds)
```

### Dispersion estimates plot

```{r dispersion_plot}
plotDispEsts(dds)
```

### heat-map with top variable genes 

```{r heatmap_plot}
getHeatmap(dds=dds, topN=500, annoGroup = c('SampleGroup'))
```


## Differential testing {.tabset}

```{r de_testing, results = "asis"}
contNames <- c()
sigUpReg <- c()
sigDownReg <- c()

for( i in 1:nrow(contTab)){
  
  numerator <- contTab$numerator[i]
  denominator <- contTab$denominator[i]
  contName <- str_c( numerator, 'vs', denominator, sep = ' ')
  
  catHeader(text = contName, level = 3, is_tabset = TRUE)
  
  res <- writeAndPlotDEResults(dds = dds,
                      factorName = factorName,
                      numerator=numerator,
                      denominator=denominator,
                      gtf = gtf,
                      DeOutDir = DeOutDir,
                      pValCutoff = 0.05
                      )
  
  contNames <- c(contNames, contName)
  sigUpReg <- c(sigUpReg, 
                sum(res$log2FoldChange > 0 & 
                      res$padj < pValCutoff, na.rm=TRUE))
  sigDownReg <- c(sigDownReg,
                  sum(res$log2FoldChange < 0 & 
                        res$padj < pValCutoff, na.rm=TRUE))
  
  # MA plot
  catHeader(text = 'MA plot', level = 4)
  p_ma <- getMAplot(res = res, numerator = numerator,
          denominator=denominator,
          topN=20, genesToShow = genesToShow)
  
  print(p_ma)
  
  cat('\n\n\n**This is MA plot, on X axis is averege gene expression, on Y-axis log 2 fold chages.**\n\n\n')
  
  
  # volcano plot
  catHeader(text = 'volcano plot', level = 4)
  
  p_vol <- getVolcanoPlot(res=res, numerator = numerator,
               denominator = denominator,
               topN = 20, genesToShow = genesToShow,
               pValCutoff = pValCutoff)
  
  print(p_vol)
  
  # p-value distribution plot
  catHeader(text = 'p-value distribution plot', level = 4)
  
  p_pval <- getPvalDistriPlot(res = res,
                  numerator = numerator,
                  denominator=denominator,
                  pValCutoff = pValCutoff)
  
  print(p_pval)
  
  # Karyogram plot
  catHeader(text = 'Karyogram plot', level = 4)
  
  kPlot <- getKaryogramPlot(dds=dds,
                 numerator = numerator,
                 denominator = denominator,
                 factorName = factorName,gtf = gtf,
                 pValCutoff = pValCutoff,
                 genome = genome

                  )
  
  print(kPlot)
  
  # gene counts plot
  catHeader(text = 'Gene counts plot', level = 4)
  
  
  geneCtsPlot <- getGeneCountsPlot(dds=dds,
                  numerator = numerator, 
                  denominator = denominator,
                  factorName = factorName,
                  gtf=gtf,
                  pValCutoff=pValCutoff,
                  topN=10, 
                  genesToShow=genesToShow)
  print(geneCtsPlot)
  
  

}

deSumTab <- tibble(contrast=contNames, 
                   sigUpReg=sigUpReg, 
                   sigDownReg=sigDownReg,
                   totalSigReg= (sigUpReg + sigDownReg)
                   )
```


## DE summary table 

```{r DE_sum_tab}

deSumTab %>% 
  datatable(style='bootstrap', caption = 'Significantly regulated genes for each contrast ') %>% 
  formatStyle(
    columns = 'contrast',
    target = 'cell',
    color= 'blue',
    ontWeight='bold'
  )

```

